src/Workspaces/Core/MSBuild/MSBuild/MSBuildWorkspace.cs

```diff
@@ -263,6 +335,64 @@ private static string GetAbsolutePath(string path, string baseDirectoryPath)
             return projectResult;
         }
 
+        /// <summary>
+        /// Reload the given projects based on their contents on disk.
+        /// </summary>
+        /// <param name="projectIds">The <see cref="ProjectId" /> set to be reloaded</param>
+        /// <param name="targets">The set of targets to execute on all projects in the solution</param>
+        /// <param name="msbuildLogger">An optional <see cref="ILogger"/> that will log msbuild results..</param>
+        /// <param name="progress">An optional <see cref="IProgress{T}"/> that will receive updates as the project is opened.</param>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> to allow cancellation of this operation.</param>
+        public async Task<ImmutableArray<Project>> ReloadProjectsAsync(
+            ImmutableArray<ProjectId> projectIds,
+            ImmutableArray<string> targets = default,
+            ILogger? msbuildLogger = null,
+            IProgress<ProjectLoadProgress>? progress = null,
+            CancellationToken cancellationToken = default)
+        {
+            if (targets == default)
+            {
+                targets = Targets.Default;
+            }
+
+            using var _1 = ArrayBuilder<Project>.GetInstance(out var projectsToReload);
+            foreach (var projectId in projectIds)
+            {
+                this.CheckProjectIsInCurrentSolution(projectId);
+                var projectResult = this.CurrentSolution.GetProject(projectId);
+                RoslynDebug.AssertNotNull(projectResult);
+                projectsToReload.Add(projectResult);
+            }
+
+            var projectMap = ProjectMap.Create(this.CurrentSolution);
+            var projectFilePaths = projectsToReload.SelectAsArray(p => p.FilePath);
+            // None of the project file paths should be null
+            var projectInfos = await _loader.LoadProjectInfoAsync(projectFilePaths!, targets, projectMap, progress, msbuildLogger, cancellationToken).ConfigureAwait(false);
+
+            var updatedProjects = await UpdateProjectsAsync(projectInfos).ConfigureAwait(false);
+
+            UpdateReferencesAfterAdd();
```

Comment:
So in the VS workspace the metadata to project reference conversion is much trickier -- if the reload means the assembly name changed for example, we may have to convert things back to metadata references. Do we need work for that?